### PR TITLE
[feature] #2040: Add integration test with transaction execution limit

### LINKED ISSUE #2040: Implement integration test with trasaction execution count checking
In this issue should be implementd the 3'rd test-case from #1890 :

> Precommit trigger meaning quantitative restriction for operation

This means that some account can perform some action only limited number of times.
It should be done with roles, not triggers. Just need to create new permission check.

### PR DESCRIPTION
**Pull request by Arjentix**
### Description of the Change

- Added new permission validator that checks for `Transfer` execution count and sets some limit per time period. I.e. it's possible to limit someone to **10** `Transfer` transactions per day
- Added test which demostrates it. Also there are additional comments so that someone from Orillion project could read this, cause this test is needed for their use-cases

### Issue

Closes #2040 

### Benefits

One more permission validator and one more integration test

### Possible Drawbacks

- I'm not sure if I used permission validators system on the right way
- Should this new validator participate in `default_permissions()`?

### CODE REVIEW

**Comment by appetrosyan** [comment]
Re: should this be part of default permissions

I think no.

File: client/tests/integration/roles.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    let alice_id = <Account as Identifiable>::Id::from_str("alice@wonderland")?;
+    let bob_id = <Account as Identifiable>::Id::from_str("bob@wonderland")?;
+    let rose_definition_id = <AssetDefinition as Identifiable>::Id::from_str("rose#wonderland")?;
+    let alice_rose_id =
+        <Asset as Identifiable>::Id::new(rose_definition_id.clone(), alice_id.clone());
+    let bob_rose_id = <Asset as Identifiable>::Id::new(rose_definition_id, bob_id.clone());
+    let role_id = <Role as Identifiable>::Id::from_str("non_privileged_user")?;
+    let rose_value = get_asset_value(&mut test_client, alice_rose_id.clone())?;
+    assert!(rose_value > COUNT + 1);
```
Comment (appetrosyan) [inline client/tests/integration/roles.rs:35]: I'd add a comment that Alice has roses from genesis. Orillion devs might not know that.

File: client/tests/integration/roles.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Registering Bob
+    let register_bob = RegisterBox::new(Account::new(bob_id, []));
+    test_client.submit_blocking(register_bob)?;
+
+    // Registering new role which sets `Transfer` execution count limit to
+    // `COUNT` for every `PERIOD_MS` milliseconds
+    let permission_token = PermissionToken::new(
+        transfer::CAN_TRANSFER_ONLY_FIXED_NUMBER_OF_TIMES_PER_PERIOD.clone(),
```
Comment (appetrosyan) [inline client/tests/integration/roles.rs:44]: Not a huge fan of this API. we probably need to make it better.

File: permissions_validators/src/public_blockchain/mod.rs [hunk 3]
```diff
@@ -42,6 +42,7 @@ pub fn default_permissions<W: WorldTrait>() -> IsInstructionAllowedBoxed<W> {
     ValidatorBuilder::new()
         .with_recursive_validator(grant_instruction_validator)
         .with_recursive_validator(transfer::OnlyOwnedAssets.or(transfer::GrantedByAssetOwner))
+        .with_recursive_validator(transfer::ExecutionCountFitsInLimit)
```
Comment (appetrosyan) [inline permissions_validators/src/public_blockchain/mod.rs:45]: I think this shouldn't be the default. Restricting transfers is specific for the Orillion use-case.

File: core/src/wsv.rs [hunk 4]
```diff
@@ -418,9 +418,7 @@ impl<W: WorldTrait> WorldStateView<W> {
     /// **Locking behaviour**: Holding references to blocks stored in the blockchain can induce
     /// deadlock. This limitation is imposed by the fact that blockchain is backed by [`dashmap::DashMap`]
     #[inline]
-    pub fn blocks(
-        &self,
-    ) -> impl Iterator<Item = impl Deref<Target = VersionedCommittedBlock> + '_> {
```
Comment (mversic) [inline core/src/wsv.rs:423]: why this change?

Comment (Arjentix) [reply core/src/wsv.rs:423]: I need to use `rev()` method which is provided by `ChainIterator`

File: client/tests/integration/roles.rs (hunk 2 above)
Comment (Arjentix) [reply client/tests/integration/roles.rs:44]: Yeah, using constant strings to identify permission looks odd. Maybe we need strong typing here too? I'll create an issue for refactoring

File: data_model/src/role.rs [hunk 5]
```diff
@@ -42,8 +43,8 @@ pub struct Id {
 impl Id {
     /// Construct role id
     #[inline]
-    pub fn new(name: impl Into<Name>) -> Self {
-        Self { name: name.into() }
+    pub fn new(name: Name) -> Self {
```
Comment (mversic) [inline data_model/src/role.rs:46]: can be `const`

File: data_model/src/role.rs [hunk 6]
```diff
@@ -81,11 +92,11 @@ impl Role {
     /// Constructor.
     #[inline]
     pub fn new(
-        id: impl Into<Id>,
+        id: <Self as Identifiable>::Id,
```
Comment (mversic) [inline data_model/src/role.rs:95]: if you want to make this change, I think there are other places in `data_model` where the same form of constructor was used

File: core/src/wsv.rs (hunk 4 above)
Comment (appetrosyan) [reply core/src/wsv.rs:423]: I think that an opaque iterator here only makes things harder.

File: permissions_validators/src/public_blockchain/transfer.rs [hunk 7]
```diff
@@ -103,3 +115,133 @@ impl<W: WorldTrait> IsGrantAllowed<W> for GrantMyAssetAccess {
...
+        )),
+    }
+}
+
+/// Find number of `Transfer` execution count which happened in last `period`
+fn count_executions<W: WorldTrait>(
+    wsv: &WorldStateView<W>,
+    authority: &AccountId,
+    period: Duration,
+) -> u32 {
```
Comment (mversic) [inline permissions_validators/src/public_blockchain/transfer.rs:216]: It makes more sense that you return `usize` here and handle conversion at the call site if needed

File: permissions_validators/src/public_blockchain/transfer.rs [hunk 8]
```diff
@@ -103,3 +115,133 @@ impl<W: WorldTrait> IsGrantAllowed<W> for GrantMyAssetAccess {
...
+    {
+        Value::U32(count) => Ok(*count),
+        _ => Err(format!(
+            "`{}` parameter has wrong value type. Expected `u32`",
+            *COUNT_PARAM_NAME
+        )),
+    }
+}
+
+/// Find number of `Transfer` execution count which happened in last `period`
```
Comment (mversic) [inline permissions_validators/src/public_blockchain/transfer.rs:211]:
```suggestion
/// Counts the number of `Transfer`s  which happened in the last `period`
```

File: client/tests/integration/roles.rs (hunk 2 above)
Comment (Arjentix) [reply client/tests/integration/roles.rs:48]: Created #2052

File: permissions_validators/src/public_blockchain/transfer.rs [hunk 9]
```diff
@@ -103,3 +115,133 @@ impl<W: WorldTrait> IsGrantAllowed<W> for GrantMyAssetAccess {
...
+fn retrieve_period(params: &BTreeMap<Name, Value>) -> Result<Duration, DenialReason> {
+    match params
+        .get(&*PERIOD_PARAM_NAME)
+        .ok_or_else(|| format!("Expected `{}` parameter", *PERIOD_PARAM_NAME))?
+    {
+        Value::U128(period) => Ok(Duration::from_millis(
+            u64::try_from(*period).map_err(|e| e.to_string())?,
+        )),
+        _ => Err(format!(
+            "`{}` parameter has wrong value type. Expected `u128`",
```
Comment (mversic) [inline permissions_validators/src/public_blockchain/transfer.rs:187]: we won't allow specifying period in other integer types?

Comment (mversic) [reply permissions_validators/src/public_blockchain/transfer.rs:187]: and what's the point of using `u128` if we're going to convert to `u64`?

Comment (mversic) [reply permissions_validators/src/public_blockchain/transfer.rs:187]: also, if this is going to remain like this, I think `Error` section in docs should include integer conversion error

File: permissions_validators/src/public_blockchain/transfer.rs [hunk 10]
```diff
@@ -103,3 +115,133 @@ impl<W: WorldTrait> IsGrantAllowed<W> for GrantMyAssetAccess {
...
+
+/// Retrieve permission parameters for `ExecutionCountFitsInLimit` validator.
+/// Returns Some if there is a suitable token and None if not
+///
+/// # Errors
+/// - Account doesn't exist
+fn retrieve_permission_params<W: WorldTrait>(
+    wsv: &WorldStateView<W>,
+    authority: &AccountId,
+) -> Result<Option<BTreeMap<Name, Value>>, DenialReason> {
```
Comment (mversic) [inline permissions_validators/src/public_blockchain/transfer.rs:162]: can we return empty collection instead of `None`?

**Review by appetrosyan — APPROVED** [review]

File: core/src/wsv.rs (hunk 4 above)
Comment (mversic) [reply core/src/wsv.rs:423]: `DoubleEndedIterator` can be used, but if you maintain that it reduces readability resolve this discussion

File: data_model/src/role.rs (hunk 5 above)
Comment (Arjentix) [reply data_model/src/role.rs:46]: Can't cause of `permissions` parameter:

> trait bounds other than `Sized` on const fn parameters are unstable

File: data_model/src/role.rs (hunk 6 above)
Comment (Arjentix) [reply data_model/src/role.rs:95]: I did it so it can be consistent with other sturctures like `Accoount`, `Domain` and etc.
Also there seems to be nothing `role::Id` can be constructed with using `From::from()` except itself.

I can't find anything like that in `data_model`. I think you eliminated such thing in your refactoring PR

File: permissions_validators/src/public_blockchain/transfer.rs (hunk 9 above)
Comment (Arjentix) [reply permissions_validators/src/public_blockchain/transfer.rs:187]: > we won't allow specifying period in other integer types?

Yeah. I think other integer types won't add more flexebility and it would be just two different ways to achieve the same result and I'm not a fan of it

> and what's the point of using `u128` if we're going to convert to `u64`?

The reason is that we don't have `u64` in `Value` and I'm not sure if we should add it just for this test case.
Another reason is that all milliseconds in Iroha 2 have `u128` type and we often convert them to `u64` in order to pass it inside `Duration`.
I think initialy the problem is [here](https://github.com/rust-lang/rust/issues/58580)

**Comment by Arjentix** [comment]
I'v just updated it to new `iroha2-dev`

**Review by appetrosyan — APPROVED** [review]

**Review by mversic — APPROVED** [review]
